### PR TITLE
Add experimental bulkpatch workflow

### PR DESCRIPTION
**Pull request by mattlwilliams**
Adds new 'bulkpatch' command designed for CI scenarios where multiple diffs may need to be applied quickly.

### CODE REVIEW

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+  // See ArcanistPatchWorkflow.php
+  private function getCommitMessage(ArcanistBundle $bundle) {
+    $revision_id    = $bundle->getRevisionID();
+    $commit_message = null;
+    $prompt_message = null;
+
+    // if we have a revision id the commit message is in differential
+
+    // TODO: See T848 for the authenticated stuff.
+    if ($revision_id && $this->isConduitAuthenticated()) {
```
Comment (artms) [inline src/workflow/ArcanistBulkPatchWorkflow.php:344]: This and second conditional statement bodies are identical? Did you intend that? Doesn't seem that second conditional statement body would yield any other results

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+      $conduit        = $this->getConduit();
+      $commit_message = $conduit->callMethodSynchronous(
+        'differential.getcommitmessage',
+        array(
+          'revision_id' => $revision_id,
+        ));
+      $prompt_message = pht(
```
Comment (artms) [inline src/workflow/ArcanistBulkPatchWorkflow.php:352]: Looks like unused

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+    if (!$commit_message && $revision_id) {
+      $this->authenticateConduit();
+      $conduit        = $this->getConduit();
+      $commit_message = $conduit->callMethodSynchronous(
+        'differential.getcommitmessage',
+        array(
+          'revision_id' => $revision_id,
+        ));
+      $prompt_message = pht(
```
Comment (artms) [inline src/workflow/ArcanistBulkPatchWorkflow.php:366]: Looks like unused

File: src/workflow/ArcanistBulkPatchWorkflow.php (hunk 1 above)
Comment (artms) [reply src/workflow/ArcanistBulkPatchWorkflow.php:344]: Ok, I see that it was not very optimal PR  done by SQ team couple of years ago, we can probably just refactor it into
```
if ($revision_id) {
   $this->authenticateConduit(); // already checks if authenticated
   $conduit        = $this->getConduit();
   $commit_message = $conduit->callMethodSynchronous(
      'differential.getcommitmessage',
      array(
        'revision_id' => $revision_id,
      ));
}
```
We might even not check if revision_id is empty, if it is empty it most likely will provide a better debugging information (like json invalid or revision_id is empty from phabricator side) than error you return later when you check if commit_message is not empty

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+      if (!strlen(trim($commit_message))) {
+        throw new ArcanistUserAbortException();
+      }
+    }
+    */
+
+    return $commit_message;
+  }
+}
+// UBER CODE
```
Comment (artms) [inline src/workflow/ArcanistBulkPatchWorkflow.php:403]: No need for this :) Whole file is Uber's :) We add these comments when we modify upstream files :)

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 5]
```diff
@@ -0,0 +1,403 @@
+<?php
+// UBER CODE
```
Comment (artms) [inline src/workflow/ArcanistBulkPatchWorkflow.php:2]: No need for this, whole file is Uber's, we're not modifying upstream file

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+        pht('Unable to apply git generated patch - %s does not apply to current working state.', $diffId));
+    }
+    return $err;
+  }
+
+  // See ArcanistPatchWorkflow.php
+  private function getCommitMessage(ArcanistBundle $bundle) {
+    $revision_id    = $bundle->getRevisionID();
+    $commit_message = null;
+    $prompt_message = null;
```
Comment (artms) [inline src/workflow/ArcanistBulkPatchWorkflow.php:339]: not actually needed

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+    $diffPatchMap = array();
+    try {
+      // Clone local repo bare
+      $err = $repository_api->execPassthru('clone --bare --local . %s', $repoPath);
+      if ($err) {
+        throw new Exception(pht("Failed to create local clone with error code: %s", $err));
+      }
+
+      // Fetch refSpec
+      $err = $repository_api->execPassthru('-C %s fetch --depth=1 --no-tags %s ' . $refSpec, $repoPath, $staging_uri);
```
[the remark this answers is not in the document]
Comment (mattlwilliams) [reply src/workflow/ArcanistBulkPatchWorkflow.php:258]: My understanding is that depth=1 improves performance (from ~60s to 5-10s), but has the unfortunate side-effect of occasionally shallowing the repo if we fetch a ref pointing to a commit already in the local repo. We avoid the shallowing effect here by creating the --bare --local clone and fetching into that instead.

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+        $err = $this->applyDiffWithFallback($diffId);
+        if ($err) {
+          return $err;
+        }
+        $this->writeOkay(
+          pht('APPLIED'),
+          pht('Successfully applied diff %s.', $diffId));
+      }
+
+      $repository_api->execPassthru('submodule update --init --recursive');
```
[the remark this answers is not in the document]
Comment (mattlwilliams) [reply src/workflow/ArcanistBulkPatchWorkflow.php:179]: I don't know with certainty whether the repositories I'm targeting will never have submodules - can revisit if this becomes problematic

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+        $err = $this->applyDiffWithFallback($diffId);
+        if ($err) {
+          return $err;
+        }
+
+        $this->writeOkay(
+          pht('APPLIED'),
+          pht('Successfully applied diff %s.', $diffId));
+
+        $repository_api->execPassthru('submodule update --init --recursive');
```
[the remark this answers is not in the document]
Comment (mattlwilliams) [reply src/workflow/ArcanistBulkPatchWorkflow.php:206]: Based on https://secure.phabricator.com/rARC73804f003972ead1bc38bfac8f146b42bfee20a7 I think we may need to execute before every commit

File: src/workflow/ArcanistBulkPatchWorkflow.php (hunk 1 above)
Comment (mattlwilliams) [reply src/workflow/ArcanistBulkPatchWorkflow.php]: Refactored getCommitMessage - had to keep revision_id check, as differential.getcommitmessage returns a template message if revision_id is null

File: src/workflow/ArcanistBulkPatchWorkflow.php [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+    $repository_api = $this->getRepositoryAPI();
+    if (!$repository_api instanceof ArcanistGitAPI) {
+      $message = pht(
+        "arc bulkpatch only works for git repositories");
+      throw new ArcanistUsageException($message);
+    }
+
+    // Eagerly load diffs to check valid args passed
+    $this->authenticateConduit();
+    foreach ($this->allDiffIds as $diffId) {
```
[the remark this answers is not in the document]
Comment (mattlwilliams) [reply src/workflow/ArcanistBulkPatchWorkflow.php:151]: nice suggestion, thanks! added a TODO for after production is in a good state

File: src/workflow/ArcanistBulkPatchWorkflow.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (artms) [reply src/workflow/ArcanistBulkPatchWorkflow.php:258]: Well, when I tested few time I've never seen 60s fetch when no depth was used, so if you're OK with this place yourself then I'm ok as well.

**Review by artms — APPROVED** [review]

**Pull request merged**